File: app/src/test/java/activity/amigosecreto/db/ParticipanteModelTest.java
package activity.amigosecreto.db;

import org.junit.Test;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.Arrays;

import static org.junit.Assert.*;

public class ParticipanteModelTest {

    @Test
    public void construtor_padrao_amigo_sorteado_e_null() {
        Participante p = new Participante();
        assertNull(p.getAmigoSorteadoId());
    }

    @Test
    public void construtor_padrao_ids_excluidos_e_lista_vazia() {
        Participante p = new Participante();
        assertNotNull(p.getIdsExcluidos());
        assertTrue(p.getIdsExcluidos().isEmpty());
    }

    @Test
    public void construtor_padrao_enviado_e_false() {
        Participante p = new Participante();
        assertFalse(p.isEnviado());
    }

    @Test
    public void setNome_getNome_round_trip() {
        Participante p = new Participante();
        p.setNome("Maria");
        assertEquals("Maria", p.getNome());
    }

    @Test
    public void setEmail_getEmail_round_trip() {
        Participante p = new Participante();
        p.setEmail("maria@email.com");
        assertEquals("maria@email.com", p.getEmail());
    }

    @Test
    public void setTelefone_getTelefone_round_trip() {
        Participante p = new Participante();
        p.setTelefone("(11) 91234-5678");
        assertEquals("(11) 91234-5678", p.getTelefone());
    }

    @Test
    public void setAmigoSorteadoId_getAmigoSorteadoId_round_trip() {
        Participante p = new Participante();
        p.setAmigoSorteadoId(99);
        assertEquals(Integer.valueOf(99), p.getAmigoSorteadoId());
    }

    @Test
    public void setEnviado_true_getEnviado_retorna_true() {
        Participante p = new Participante();
        p.setEnviado(true);
        assertTrue(p.isEnviado());
    }

    @Test
    public void setIdsExcluidos_getIdsExcluidos_round_trip() {
        Participante p = new Participante();
        p.setIdsExcluidos(Arrays.asList(1, 2, 3));
        assertEquals(3, p.getIdsExcluidos().size());
        assertTrue(p.getIdsExcluidos().contains(2));
    }

    @Test
    public void toString_retorna_nome() {
        Participante p = new Participante();
        p.setNome("Joao");
        assertEquals("Joao", p.toString());
    }

    @Test
    public void toString_com_nome_null_retorna_string_vazia() {
        Participante p = new Participante();
        assertEquals("", p.toString());
    }

    @Test
    public void serializable_round_trip_preserva_campos_principais() throws Exception {
        Participante original = new Participante();
        original.setId(10);
        original.setNome("Carlos");
        original.setEmail("carlos@teste.com");
        original.setTelefone("(21) 99999-8888");
        original.setEnviado(true);
        original.setAmigoSorteadoId(5);

        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        ObjectOutputStream oos = new ObjectOutputStream(baos);
        oos.writeObject(original);
        oos.close();

        ByteArrayInputStream bais = new ByteArrayInputStream(baos.toByteArray());
        ObjectInputStream ois = new ObjectInputStream(bais);
        Participante desserializado = (Participante) ois.readObject();
        ois.close();

        assertEquals(original.getId(), desserializado.getId());
        assertEquals(original.getNome(), desserializado.getNome());
        assertEquals(original.getEmail(), desserializado.getEmail());
        assertEquals(original.getTelefone(), desserializado.getTelefone());
        assertEquals(original.isEnviado(), desserializado.isEnviado());
        assertEquals(original.getAmigoSorteadoId(), desserializado.getAmigoSorteadoId());
    }
}

File: app/src/test/java/activity/amigosecreto/db/ParticipanteModelTest.kt
package activity.amigosecreto.db

import org.junit.Assert.*
import org.junit.Test
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream

class ParticipanteModelTest {

    @Test
    fun construtor_padrao_amigo_sorteado_e_null() {
        assertNull(Participante().amigoSorteadoId)
    }

    @Test
    fun construtor_padrao_ids_excluidos_e_lista_vazia() {
        val p = Participante()
        assertNotNull(p.idsExcluidos)
        assertTrue(p.idsExcluidos.isEmpty())
    }

    @Test
    fun construtor_padrao_enviado_e_false() {
        assertFalse(Participante().isEnviado)
    }

    @Test
    fun setNome_getNome_round_trip() {
        val p = Participante()
        p.nome = "Maria"
        assertEquals("Maria", p.nome)
    }

    @Test
    fun setEmail_getEmail_round_trip() {
        val p = Participante()
        p.email = "maria@email.com"
        assertEquals("maria@email.com", p.email)
    }

    @Test
    fun setTelefone_getTelefone_round_trip() {
        val p = Participante()
        p.telefone = "(11) 91234-5678"
        assertEquals("(11) 91234-5678", p.telefone)
    }

    @Test
    fun setAmigoSorteadoId_getAmigoSorteadoId_round_trip() {
        val p = Participante()
        p.amigoSorteadoId = 99
        assertEquals(99, p.amigoSorteadoId)
    }

    @Test
    fun setEnviado_true_getEnviado_retorna_true() {
        val p = Participante()
        p.isEnviado = true
        assertTrue(p.isEnviado)
    }

    @Test
    fun setIdsExcluidos_getIdsExcluidos_round_trip() {
        val p = Participante()
        p.idsExcluidos = mutableListOf(1, 2, 3)
        assertEquals(3, p.idsExcluidos.size)
        assertTrue(p.idsExcluidos.contains(2))
    }

    @Test
    fun setIdsExcluidos_comArraysAsList_continuaMutavel() {
        // setter deve fazer cópia defensiva para que add() não lance UnsupportedOperationException
        val p = Participante()
        p.idsExcluidos = mutableListOf(1, 2, 3)
        p.idsExcluidos.add(4)
        assertEquals(4, p.idsExcluidos.size)
        assertTrue(p.idsExcluidos.contains(4))
    }

    @Test
    fun toString_retorna_nome() {
        val p = Participante()
        p.nome = "Joao"
        assertEquals("Joao", p.toString())
    }

    @Test
    fun toString_com_nome_null_retorna_string_vazia() {
        assertEquals("", Participante().toString())
    }

    @Test
    fun serializable_round_trip_preserva_campos_principais() {
        val original = Participante().apply {
            id = 10
            nome = "Carlos"
            email = "carlos@teste.com"
            telefone = "(21) 99999-8888"
            isEnviado = true
            amigoSorteadoId = 5
        }

        val baos = ByteArrayOutputStream()
        ObjectOutputStream(baos).use { it.writeObject(original) }

        val desserializado = ObjectInputStream(ByteArrayInputStream(baos.toByteArray())).use {
            it.readObject() as Participante
        }

        assertEquals(original.id, desserializado.id)
        assertEquals(original.nome, desserializado.nome)
        assertEquals(original.email, desserializado.email)
        assertEquals(original.telefone, desserializado.telefone)
        assertEquals(original.isEnviado, desserializado.isEnviado)
        assertEquals(original.amigoSorteadoId, desserializado.amigoSorteadoId)
    }
}